### DISCUSSION TITLE
Include custom html

### DISCUSSION THREAD

**Discussion by shapeshifta78** [remix-run/remix#2680] [Q&A]
Hi, how can I include custom html in root.tsx? We have some generated html pages from another team which we would like to include as the project root. Is that possible? A full html file with head (with meta tags and so on) and body (with html elements)

**Comment by sergiodxa** [comment]
You can render html directly in react using the dangerouslySetInnerHtml prop in any element.

**Reply by shapeshifta78** [reply]
yes, I am aware of that, but it is a full html file with head (with meta tags and so on) and body (with html elements). How would that be possible with remix? At the moment we use it as an entry for our react application.

**Reply by sergiodxa** [reply]
It's not possible, but if you know the URL of that HTML, you could use a resource route to send the HTML directly.

```ts
import { html } from "remix-utils";

export let loader: LoaderFunction = async () => {
  return html(await getHTML());
}
```

**Reply by shapeshifta78** [reply]
Thanks and would it be possible to integrate content somehow? Maybe by splitting the file or something?

**Reply by sergiodxa** [reply]
What do you mean by integrate content?

**Reply by shapeshifta78** [reply]
In root.tsx there is the main route which includes head and body of the project. I need to merge remix components with the custom html file.

 I tried similar things with nextjs via splitting the html content and using innerhtml, but next was complaining because of the missing head component. I was hoping to find a better way with remix. 

I can give a more elaborate example tomorrow if it's not clear enough. Thanks for taking the time 🙂

**Comment by KieranGreenwood** [comment]
Did you find a solution to this in the end, I am trying to do the same thing?

template is a object containing a bunch of block html strings that I am trying to integrate with remix functionality.

```
export default function FrontEnd() {
	const { template } = useLoaderData();

	return (
		<>
			<html>
				<head dangerouslySetInnerHTML={{ __html: template.Head }} />
				{/* <Meta /> */}
				{/* <Links /> */}
				<body>
					<header dangerouslySetInnerHTML={{ __html: template.Header }} />
					<main>
						<Outlet />
					</main>
					<footer dangerouslySetInnerHTML={{ __html: template.Footer }} />
					{/* <div dangerouslySetInnerHTML={{ __html: template.FootScripts }} /> */}
					<Scripts />
				</body>
			</html>
		</>
	);
}
```

It mostly works but because i can't find a way to include `<Meta />` and `<Links />` in the head alongside my dynamic html, I get hydration errors and flickering rendering.

I have tried wrapping it all in a normal head and using `<React.Fragment dangerouslySetInnerHtml />` and also `<div danger..>` but Fragment doesnt support the prop and div doesnt work in the head.

**Reply by KieranGreenwood** [reply]
Found a way using [html-react-parser](https://www.npmjs.com/package/html-react-parser).

```
<head>
	{parse(template.Head)}
	<Meta />
	<Links />
</head>
```

**Comment by shapeshifta78** [comment]
No solution, sorry. We don't use remix

KieranGreenwood ***@***.***> schrieb am Fr., 17. Mai 2024,
18:50:
